File: examples/python/dex_examples/patterns/sequentially-chunked-attribute-map/chunked_subscriber_flow.py
# Copyright (c) 2022-2026 Super Durable, Inc.
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.

from __future__ import annotations

from dataclasses import dataclass

from dex import Attribute, AttributeMap, Context, Flow, PersistenceSchema, RPCResult, StepList, rpc

FLOW_ID = "chunked-subscriber-archive"
CURRENT_CHUNK_INSTANCE = "current"
SUBSCRIBER_CHUNK_SIZE = 100


@dataclass
class Subscriber:
    sequence: int
    subscriber_id: str
    delivery_address: str


@dataclass
class SubscriberChunk:
    subscribers: list[Subscriber]


@dataclass
class SubscriberArchiveState:
    next_sequence: int
    latest_archived_chunk_token: str = ""


@dataclass
class RegisterSubscriberInput:
    subscriber_id: str
    delivery_address: str


@dataclass
class GetSubscriberPageInput:
    page_token: str


@dataclass
class SubscriberPage:
    page_token: str
    subscribers: list[Subscriber]
    next_page_token: str = ""


class ChunkedSubscriberFlow(Flow[None]):
    archive_state = Attribute("subscriber_archive_state", SubscriberArchiveState)
    subscriber_chunks = AttributeMap("subscriber_chunks", SubscriberChunk)

    def get_steps(self) -> StepList[None]:
        return StepList.empty()

    def get_persistence_schema(self) -> PersistenceSchema:
        return PersistenceSchema.of(self.archive_state, self.subscriber_chunks)

    @rpc(
        name="RegisterSubscriber",
        lock_attributes=(archive_state.lock(), subscriber_chunks.lock(CURRENT_CHUNK_INSTANCE)),
        load_attribute_map_instances=(subscriber_chunks.load(CURRENT_CHUNK_INSTANCE),),
    )
    def register_subscriber(
        self,
        context: Context,
        input: RegisterSubscriberInput,
    ) -> RPCResult[Subscriber]:
        if not input.subscriber_id:
            raise ValueError("subscriberId is required")
        if not input.delivery_address:
            raise ValueError("deliveryAddress is required")

        archive_state = self.archive_state.get(context)
        current_chunk = self.subscriber_chunks.get(context, CURRENT_CHUNK_INSTANCE)
        subscriber = Subscriber(
            archive_state.next_sequence,
            input.subscriber_id,
            input.delivery_address,
        )
        if len(current_chunk.subscribers) == SUBSCRIBER_CHUNK_SIZE:
            archive_token = archive_chunk_token(current_chunk.subscribers[0].sequence)
            self.subscriber_chunks.set(context, archive_token, current_chunk)
            archive_state.latest_archived_chunk_token = archive_token
            current_chunk = SubscriberChunk([subscriber])
        else:
            current_chunk.subscribers.append(subscriber)
        archive_state.next_sequence += 1
        self.subscriber_chunks.set(context, CURRENT_CHUNK_INSTANCE, current_chunk)
        self.archive_state.set(context, archive_state)
        return RPCResult(subscriber)

    @rpc(name="GetSubscriberPage")
    def get_subscriber_page(
        self,
        context: Context,
        input: GetSubscriberPageInput,
    ) -> RPCResult[SubscriberPage]:
        page_token = validate_subscriber_page_token(input.page_token)
        try:
            chunk = self.subscriber_chunks.get(context, page_token)
        except KeyError as error:
            raise ValueError(f'subscriber page "{page_token}" not found') from error
        page = SubscriberPage(page_token, list(reversed(chunk.subscribers)))
        if not chunk.subscribers:
            return RPCResult(page)
        if page_token == CURRENT_CHUNK_INSTANCE:
            page.next_page_token = self.archive_state.get(
                context
            ).latest_archived_chunk_token
        else:
            first_sequence = int(page_token)
            if first_sequence > 1:
                page.next_page_token = archive_chunk_token(
                    first_sequence - SUBSCRIBER_CHUNK_SIZE
                )
        return RPCResult(page)


def validate_subscriber_page_token(page_token: str) -> str:
    if not page_token:
        return CURRENT_CHUNK_INSTANCE
    if page_token == CURRENT_CHUNK_INSTANCE:
        return page_token
    if len(page_token) != 20 or not page_token.isascii() or not page_token.isdigit():
        raise ValueError(f'invalid page token "{page_token}"')
    first_sequence = int(page_token)
    if first_sequence == 0 or (first_sequence - 1) % SUBSCRIBER_CHUNK_SIZE != 0:
        raise ValueError(f'invalid page token "{page_token}"')
    return page_token


def archive_chunk_token(first_sequence: int) -> str:
    return f"{first_sequence:020d}"

File: examples/typescript/src/patterns/sequentially-chunked-attribute-map/chunked-subscriber-flow.ts
/*
 * Copyright (c) 2022-2026 Super Durable, Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 *     http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */

import {
  Attribute,
  AttributeMap,
  StepList,
  jsonCodec,
  rpc,
  type Context,
  type Flow,
  type PersistenceSchema,
  type RPCResult,
} from "@superdurable/dex";

export const CHUNKED_SUBSCRIBER_FLOW_ID = "chunked-subscriber-archive";
export const CURRENT_CHUNK_INSTANCE = "current";
export const SUBSCRIBER_CHUNK_SIZE = 100;

export interface Subscriber {
  readonly sequence: number;
  readonly subscriberId: string;
  readonly deliveryAddress: string;
}

export interface SubscriberChunk {
  readonly subscribers: readonly Subscriber[];
}

export interface SubscriberArchiveState {
  readonly nextSequence: number;
  readonly latestArchivedChunkToken: string;
}

export interface RegisterSubscriberInput {
  readonly subscriberId: string;
  readonly deliveryAddress: string;
}

export interface GetSubscriberPageInput {
  readonly pageToken: string;
}

export interface SubscriberPage {
  readonly pageToken: string;
  readonly subscribers: readonly Subscriber[];
  readonly nextPageToken: string;
}

const subscriberCodec = jsonCodec<Subscriber>();
const registerSubscriberInputCodec = jsonCodec<RegisterSubscriberInput>();
const subscriberPageInputCodec = jsonCodec<GetSubscriberPageInput>();
const subscriberPageCodec = jsonCodec<SubscriberPage>();
const archiveState = new Attribute(
  "subscriber_archive_state",
  jsonCodec<SubscriberArchiveState>(),
);
const subscriberChunks = new AttributeMap(
  "subscriber_chunks",
  jsonCodec<SubscriberChunk>(),
);

export class ChunkedSubscriberFlow implements Flow<void> {
  public readonly archiveState = archiveState;
  public readonly subscriberChunks = subscriberChunks;

  public getFlowType(): string {
    return "ChunkedSubscriberFlow";
  }

  public getSteps() {
    return StepList.empty();
  }

  public getPersistenceSchema(): PersistenceSchema {
    return { attributes: [this.archiveState, this.subscriberChunks] };
  }

  @rpc({
    name: "RegisterSubscriber",
    inputCodec: registerSubscriberInputCodec,
    outputCodec: subscriberCodec,
    lockAttributes: [archiveState.lock(), subscriberChunks.lock(CURRENT_CHUNK_INSTANCE)],
    loadAttributeMapInstances: [subscriberChunks.load(CURRENT_CHUNK_INSTANCE)],
  })
  public registerSubscriber(
    context: Context,
    input: RegisterSubscriberInput,
  ): RPCResult<Subscriber> {
    if (input.subscriberId.length === 0) {
      throw new Error("subscriberId is required");
    }
    if (input.deliveryAddress.length === 0) {
      throw new Error("deliveryAddress is required");
    }
    const state = this.archiveState.get(context);
    const currentChunk = this.subscriberChunks.get(context, CURRENT_CHUNK_INSTANCE);
    const subscriber: Subscriber = {
      sequence: state.nextSequence,
      subscriberId: input.subscriberId,
      deliveryAddress: input.deliveryAddress,
    };
    let nextChunk: SubscriberChunk;
    let latestArchivedChunkToken = state.latestArchivedChunkToken;
    if (currentChunk.subscribers.length === SUBSCRIBER_CHUNK_SIZE) {
      latestArchivedChunkToken = archiveChunkToken(currentChunk.subscribers[0]!.sequence);
      this.subscriberChunks.set(context, latestArchivedChunkToken, currentChunk);
      nextChunk = { subscribers: [subscriber] };
    } else {
      nextChunk = { subscribers: [...currentChunk.subscribers, subscriber] };
    }
    this.subscriberChunks.set(context, CURRENT_CHUNK_INSTANCE, nextChunk);
    this.archiveState.set(context, {
      nextSequence: state.nextSequence + 1,
      latestArchivedChunkToken,
    });
    return { output: subscriber };
  }

  @rpc({
    name: "GetSubscriberPage",
    inputCodec: subscriberPageInputCodec,
    outputCodec: subscriberPageCodec,
  })
  public getSubscriberPage(
    context: Context,
    input: GetSubscriberPageInput,
  ): RPCResult<SubscriberPage> {
    const pageToken = validateSubscriberPageToken(input.pageToken);
    const chunk = this.subscriberChunks.get(context, pageToken);
    if (chunk === undefined) {
      throw new Error(`subscriber page "${pageToken}" not found`);
    }
    let nextPageToken = "";
    if (chunk.subscribers.length > 0) {
      if (pageToken === CURRENT_CHUNK_INSTANCE) {
        nextPageToken = this.archiveState.get(context).latestArchivedChunkToken;
      } else {
        const firstSequence = Number(pageToken);
        if (firstSequence > 1) {
          nextPageToken = archiveChunkToken(firstSequence - SUBSCRIBER_CHUNK_SIZE);
        }
      }
    }
    return {
      output: {
        pageToken,
        subscribers: [...chunk.subscribers].reverse(),
        nextPageToken,
      },
    };
  }
}

export function validateSubscriberPageToken(pageToken: string): string {
  if (pageToken.length === 0) {
    return CURRENT_CHUNK_INSTANCE;
  }
  if (pageToken === CURRENT_CHUNK_INSTANCE) {
    return pageToken;
  }
  if (!/^\d{20}$/.test(pageToken)) {
    throw new Error(`invalid page token "${pageToken}"`);
  }
  const firstSequence = Number(pageToken);
  if (firstSequence === 0 || (firstSequence - 1) % SUBSCRIBER_CHUNK_SIZE !== 0) {
    throw new Error(`invalid page token "${pageToken}"`);
  }
  return pageToken;
}

export function archiveChunkToken(firstSequence: number): string {
  return String(firstSequence).padStart(20, "0");
}

export const chunkedSubscriberFlow = new ChunkedSubscriberFlow();
